feat: Support ephemeral columns

When I try to use [ephemeral columns](https://clickhouse.com/docs/sql-reference/statements/create/table#ephemeral) in ClickHouse. The create and insert works, but throws error when querying it back.

```
panic: readData: block decode: unexpected value 111 for boolean

goroutine 1 [running]:
main.main()
	/home/kavi/src/play-go2/ephemeral.go:57 +0x48b
exit status 2
```

Here is the code I used
```
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/ClickHouse/clickhouse-go/v2"
)

func main() {
	conn, err := clickhouse.Open(&clickhouse.Options{
		Protocol: clickhouse.HTTP,
		Addr:     []string{"localhost:8123"},
		Auth: clickhouse.Auth{
			Database: "default",
			Username: "default",
			Password: "",
		},
	})

	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to connect: %v\n", err)
		os.Exit(1)
	}
	defer conn.Close()

	// Execute query that will throw exception on row 3
	ctx := context.Background()
	ddl := `CREATE OR REPLACE TABLE test
(
    id UInt64,
    unhexed String EPHEMERAL,
    hexed FixedString(4) DEFAULT unhex(unhexed)
)
ENGINE = MergeTree
ORDER BY id;`
	if err := conn.Exec(ctx, ddl); err != nil {
		panic(err)
	}

	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`

	if err := conn.Exec(ctx, i); err != nil {
		panic(err)
	}

	query := `SELECT
    id,
    hexed,
    hex(hexed)
FROM test
FORMAT Vertical;`

	rows, err := conn.Query(ctx, query)
	if err != nil {
		panic(err)
	}

	for rows.Next() {
		var (
			id  uint64
			un  string
			hex string
		)
		if err := rows.Scan(&id, &un, &hex); err != nil {
			panic(err)
		}
		fmt.Println("id: ", id, "un: ", un, "hex: ", hex)
	}
}

```

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -5,7 +5,7 @@ services:
     networks:
       - clickhouse
     container_name: clickhouse
-    image: clickhouse/clickhouse-server
+    image: clickhouse/clickhouse-server:25.10
     ports:
       - 127.0.0.1:8123:8123
       - 127.0.0.1:9000:9000

--- a/examples/clickhouse_api/ephemeral_http.go
+++ b/examples/clickhouse_api/ephemeral_http.go
@@ -1,0 +1,57 @@
+package clickhouse_api
+
+import (
+	"context"
+	"fmt"
+)
+
+func EphemeralColumnHTTP() error {
+	conn, err := GetHTTPConnection("http-ephemeral-column", nil, nil, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to connect: %v\n", err)
+	}
+	defer conn.Close()
+
+	ctx := context.Background()
+	ddl := `CREATE OR REPLACE TABLE test
+(
+    id UInt64,
+    unhexed String EPHEMERAL,
+    hexed FixedString(4) DEFAULT unhex(unhexed)
+)
+ENGINE = MergeTree
+ORDER BY id;`
+	if err := conn.Exec(ctx, ddl); err != nil {
+		return err
+	}
+
+	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`
+
+	if err := conn.Exec(ctx, i); err != nil {
+		return err
+	}
+
+	query := `SELECT
+    id,
+    hexed,
+    hex(hexed)
+FROM test;`
+
+	rows, err := conn.Query(ctx, query)
+	if err != nil {
+		return err
+	}
+
+	for rows.Next() {
+		var (
+			id  uint64
+			un  string
+			hex string
+		)
+		if err := rows.Scan(&id, &un, &hex); err != nil {
+			panic(err)
+		}
+		fmt.Println("id: ", id, "un: ", un, "hex: ", hex)
+	}
+	return nil
+}

--- a/examples/clickhouse_api/ephemeral_http.go
+++ b/examples/clickhouse_api/ephemeral_http.go
@@ -13,14 +13,16 @@ func EphemeralColumnHTTP() error {
 	defer conn.Close()
 
 	ctx := context.Background()
-	ddl := `CREATE OR REPLACE TABLE test
-(
-    id UInt64,
-    unhexed String EPHEMERAL,
-    hexed FixedString(4) DEFAULT unhex(unhexed)
-)
-ENGINE = MergeTree
-ORDER BY id;`
+	ddl := `
+	CREATE OR REPLACE TABLE test
+	(
+		id UInt64,
+		unhexed String EPHEMERAL,
+		hexed FixedString(4) DEFAULT unhex(unhexed)
+	)
+	ENGINE = MergeTree
+	ORDER BY id;`
+
 	if err := conn.Exec(ctx, ddl); err != nil {
 		return err
 	}
@@ -31,11 +33,12 @@ ORDER BY id;`
 		return err
 	}
 
-	query := `SELECT
-    id,
-    hexed,
-    hex(hexed)
-FROM test;`
+	query := `
+	SELECT
+		id,
+		hexed,
+		hex(hexed)
+	FROM test;`
 
 	rows, err := conn.Query(ctx, query)
 	if err != nil {

--- a/examples/clickhouse_api/ephemeral_http.go
+++ b/examples/clickhouse_api/ephemeral_http.go
@@ -10,6 +10,8 @@ func EphemeralColumnHTTP() error {
 	if err != nil {
 		return fmt.Errorf("Failed to connect: %v\n", err)
 	}
+
+	//NOTE: never forget to close the connection at the end.
 	defer conn.Close()
 
 	ctx := context.Background()
@@ -26,12 +28,14 @@ func EphemeralColumnHTTP() error {
 	if err := conn.Exec(ctx, ddl); err != nil {
 		return err
 	}
+	fmt.Println("Created table 'test'")
 
 	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`
 
 	if err := conn.Exec(ctx, i); err != nil {
 		return err
 	}
+	fmt.Println("Inserted rows into 'test'")
 
 	query := `
 	SELECT
@@ -45,6 +49,7 @@ func EphemeralColumnHTTP() error {
 		return err
 	}
 
+	fmt.Println("Query Results:")
 	for rows.Next() {
 		var (
 			id  uint64

--- a/examples/clickhouse_api/ephemeral_native.go
+++ b/examples/clickhouse_api/ephemeral_native.go
@@ -1,0 +1,57 @@
+package clickhouse_api
+
+import (
+	"context"
+	"fmt"
+)
+
+func EphemeralColumnNative() error {
+	conn, err := GetNativeConnection(nil, nil, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to connect: %v\n", err)
+	}
+	defer conn.Close()
+
+	ctx := context.Background()
+	ddl := `CREATE OR REPLACE TABLE test
+(
+    id UInt64,
+    unhexed String EPHEMERAL,
+    hexed FixedString(4) DEFAULT unhex(unhexed)
+)
+ENGINE = MergeTree
+ORDER BY id;`
+	if err := conn.Exec(ctx, ddl); err != nil {
+		return err
+	}
+
+	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`
+
+	if err := conn.Exec(ctx, i); err != nil {
+		return err
+	}
+
+	query := `SELECT
+    id,
+    hexed,
+    hex(hexed)
+FROM test;`
+
+	rows, err := conn.Query(ctx, query)
+	if err != nil {
+		return err
+	}
+
+	for rows.Next() {
+		var (
+			id  uint64
+			un  string
+			hex string
+		)
+		if err := rows.Scan(&id, &un, &hex); err != nil {
+			panic(err)
+		}
+		fmt.Println("id: ", id, "un: ", un, "hex: ", hex)
+	}
+	return nil
+}

--- a/examples/clickhouse_api/ephemeral_native.go
+++ b/examples/clickhouse_api/ephemeral_native.go
@@ -10,6 +10,8 @@ func EphemeralColumnNative() error {
 	if err != nil {
 		return fmt.Errorf("Failed to connect: %v\n", err)
 	}
+
+	//NOTE: never forget to close the connection at the end.
 	defer conn.Close()
 
 	ctx := context.Background()
@@ -26,12 +28,14 @@ func EphemeralColumnNative() error {
 	if err := conn.Exec(ctx, ddl); err != nil {
 		return err
 	}
+	fmt.Println("Created table 'test'")
 
 	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`
 
 	if err := conn.Exec(ctx, i); err != nil {
 		return err
 	}
+	fmt.Println("Inserted rows into 'test'")
 
 	query := `
 	SELECT
@@ -45,6 +49,7 @@ func EphemeralColumnNative() error {
 		return err
 	}
 
+	fmt.Println("Query Results:")
 	for rows.Next() {
 		var (
 			id  uint64

--- a/examples/clickhouse_api/ephemeral_native.go
+++ b/examples/clickhouse_api/ephemeral_native.go
@@ -13,14 +13,16 @@ func EphemeralColumnNative() error {
 	defer conn.Close()
 
 	ctx := context.Background()
-	ddl := `CREATE OR REPLACE TABLE test
-(
-    id UInt64,
-    unhexed String EPHEMERAL,
-    hexed FixedString(4) DEFAULT unhex(unhexed)
-)
-ENGINE = MergeTree
-ORDER BY id;`
+	ddl := `
+	CREATE OR REPLACE TABLE test
+	(
+		id UInt64,
+		unhexed String EPHEMERAL,
+		hexed FixedString(4) DEFAULT unhex(unhexed)
+	)
+	ENGINE = MergeTree
+	ORDER BY id;`
+
 	if err := conn.Exec(ctx, ddl); err != nil {
 		return err
 	}
@@ -31,11 +33,12 @@ ORDER BY id;`
 		return err
 	}
 
-	query := `SELECT
-    id,
-    hexed,
-    hex(hexed)
-FROM test;`
+	query := `
+	SELECT
+		id,
+		hexed,
+		hex(hexed)
+	FROM test;`
 
 	rows, err := conn.Query(ctx, query)
 	if err != nil {

--- a/examples/clickhouse_api/main_test.go
+++ b/examples/clickhouse_api/main_test.go
@@ -60,6 +60,11 @@ func TestAsyncInsert(t *testing.T) {
 	require.NoError(t, AsyncInsertHTTP())
 }
 
+func TestEphemeralColumn(t *testing.T) {
+	require.NoError(t, EphemeralColumnNative())
+	require.NoError(t, EphemeralColumnNative())
+}
+
 func TestBatchInsert(t *testing.T) {
 	require.NoError(t, BatchInsert())
 }

--- a/examples/std/ephemeral_http.go
+++ b/examples/std/ephemeral_http.go
@@ -1,0 +1,61 @@
+package std
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+)
+
+func EphemeralColumnHTTP() error {
+	conn, err := GetStdOpenDBConnection(clickhouse.HTTP, nil, nil, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to connect: %v\n", err)
+	}
+	defer conn.Close()
+
+	ctx := context.Background()
+	ddl := `CREATE OR REPLACE TABLE test
+(
+    id UInt64,
+    unhexed String EPHEMERAL,
+    hexed FixedString(4) DEFAULT unhex(unhexed)
+)
+ENGINE = MergeTree
+ORDER BY id;`
+	_, err = conn.ExecContext(ctx, ddl)
+	if err != nil {
+		return err
+	}
+
+	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`
+
+	_, err = conn.ExecContext(ctx, i)
+	if err != nil {
+		return err
+	}
+
+	query := `SELECT
+    id,
+    hexed,
+    hex(hexed)
+FROM test;`
+
+	rows, err := conn.QueryContext(ctx, query)
+	if err != nil {
+		return err
+	}
+
+	for rows.Next() {
+		var (
+			id  uint64
+			un  string
+			hex string
+		)
+		if err := rows.Scan(&id, &un, &hex); err != nil {
+			panic(err)
+		}
+		fmt.Println("id: ", id, "un: ", un, "hex: ", hex)
+	}
+	return nil
+}

--- a/examples/std/ephemeral_http.go
+++ b/examples/std/ephemeral_http.go
@@ -15,14 +15,16 @@ func EphemeralColumnHTTP() error {
 	defer conn.Close()
 
 	ctx := context.Background()
-	ddl := `CREATE OR REPLACE TABLE test
-(
-    id UInt64,
-    unhexed String EPHEMERAL,
-    hexed FixedString(4) DEFAULT unhex(unhexed)
-)
-ENGINE = MergeTree
-ORDER BY id;`
+	ddl := `
+	CREATE OR REPLACE TABLE test
+	(
+		id UInt64,
+		unhexed String EPHEMERAL,
+		hexed FixedString(4) DEFAULT unhex(unhexed)
+	)
+	ENGINE = MergeTree
+	ORDER BY id;`
+
 	_, err = conn.ExecContext(ctx, ddl)
 	if err != nil {
 		return err
@@ -35,11 +37,12 @@ ORDER BY id;`
 		return err
 	}
 
-	query := `SELECT
-    id,
-    hexed,
-    hex(hexed)
-FROM test;`
+	query := `
+	SELECT
+		id,
+		hexed,
+		hex(hexed)
+	FROM test;`
 
 	rows, err := conn.QueryContext(ctx, query)
 	if err != nil {

--- a/examples/std/ephemeral_http.go
+++ b/examples/std/ephemeral_http.go
@@ -12,6 +12,8 @@ func EphemeralColumnHTTP() error {
 	if err != nil {
 		return fmt.Errorf("Failed to connect: %v\n", err)
 	}
+
+	//NOTE: never forget to close the connection at the end.
 	defer conn.Close()
 
 	ctx := context.Background()
@@ -29,6 +31,7 @@ func EphemeralColumnHTTP() error {
 	if err != nil {
 		return err
 	}
+	fmt.Println("Created table 'test'")
 
 	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`
 
@@ -36,6 +39,7 @@ func EphemeralColumnHTTP() error {
 	if err != nil {
 		return err
 	}
+	fmt.Println("Inserted rows into 'test'")
 
 	query := `
 	SELECT
@@ -49,6 +53,7 @@ func EphemeralColumnHTTP() error {
 		return err
 	}
 
+	fmt.Println("Query Results:")
 	for rows.Next() {
 		var (
 			id  uint64

--- a/examples/std/ephemeral_native.go
+++ b/examples/std/ephemeral_native.go
@@ -15,14 +15,16 @@ func EphemeralColumnNative() error {
 	defer conn.Close()
 
 	ctx := context.Background()
-	ddl := `CREATE OR REPLACE TABLE test
-(
-    id UInt64,
-    unhexed String EPHEMERAL,
-    hexed FixedString(4) DEFAULT unhex(unhexed)
-)
-ENGINE = MergeTree
-ORDER BY id;`
+	ddl := `
+	CREATE OR REPLACE TABLE test
+	(
+		id UInt64,
+		unhexed String EPHEMERAL,
+		hexed FixedString(4) DEFAULT unhex(unhexed)
+	)
+	ENGINE = MergeTree
+	ORDER BY id;`
+
 	_, err = conn.ExecContext(ctx, ddl)
 	if err != nil {
 		return err
@@ -35,11 +37,12 @@ ORDER BY id;`
 		return err
 	}
 
-	query := `SELECT
-    id,
-    hexed,
-    hex(hexed)
-FROM test;`
+	query := `
+	SELECT
+		id,
+		hexed,
+		hex(hexed)
+	FROM test;`
 
 	rows, err := conn.QueryContext(ctx, query)
 	if err != nil {

--- a/examples/std/ephemeral_native.go
+++ b/examples/std/ephemeral_native.go
@@ -12,6 +12,8 @@ func EphemeralColumnNative() error {
 	if err != nil {
 		return fmt.Errorf("Failed to connect: %v\n", err)
 	}
+
+	//NOTE: never forget to close the connection at the end.
 	defer conn.Close()
 
 	ctx := context.Background()
@@ -29,6 +31,7 @@ func EphemeralColumnNative() error {
 	if err != nil {
 		return err
 	}
+	fmt.Println("Created table 'test'")
 
 	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`
 
@@ -36,6 +39,7 @@ func EphemeralColumnNative() error {
 	if err != nil {
 		return err
 	}
+	fmt.Println("Inserted rows into 'test'")
 
 	query := `
 	SELECT
@@ -49,6 +53,7 @@ func EphemeralColumnNative() error {
 		return err
 	}
 
+	fmt.Println("Query Results:")
 	for rows.Next() {
 		var (
 			id  uint64

--- a/examples/std/ephemeral_native.go
+++ b/examples/std/ephemeral_native.go
@@ -1,0 +1,61 @@
+package std
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+)
+
+func EphemeralColumnNative() error {
+	conn, err := GetStdOpenDBConnection(clickhouse.Native, nil, nil, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to connect: %v\n", err)
+	}
+	defer conn.Close()
+
+	ctx := context.Background()
+	ddl := `CREATE OR REPLACE TABLE test
+(
+    id UInt64,
+    unhexed String EPHEMERAL,
+    hexed FixedString(4) DEFAULT unhex(unhexed)
+)
+ENGINE = MergeTree
+ORDER BY id;`
+	_, err = conn.ExecContext(ctx, ddl)
+	if err != nil {
+		return err
+	}
+
+	i := `INSERT INTO test (id, unhexed) VALUES (1, '5a90b714');`
+
+	_, err = conn.ExecContext(ctx, i)
+	if err != nil {
+		return err
+	}
+
+	query := `SELECT
+    id,
+    hexed,
+    hex(hexed)
+FROM test;`
+
+	rows, err := conn.QueryContext(ctx, query)
+	if err != nil {
+		return err
+	}
+
+	for rows.Next() {
+		var (
+			id  uint64
+			un  string
+			hex string
+		)
+		if err := rows.Scan(&id, &un, &hex); err != nil {
+			panic(err)
+		}
+		fmt.Println("id: ", id, "un: ", un, "hex: ", hex)
+	}
+	return nil
+}

--- a/examples/std/main_test.go
+++ b/examples/std/main_test.go
@@ -94,6 +94,11 @@ func TestStdAsyncInsert(t *testing.T) {
 	require.NoError(t, AsyncInsertHTTP_WithPrepare())
 }
 
+func TestStdEphemeralColumn(t *testing.T) {
+	require.NoError(t, EphemeralColumnNative())
+	require.NoError(t, EphemeralColumnHTTP())
+}
+
 func TestStdMapInsertRead(t *testing.T) {
 	require.NoError(t, MapInsertRead())
 }

